GOALS.md

```diff
@@ -0,0 +1,233 @@
+
+## Making a Safe library
+
+Safety is at the core of Rust's language and API design. If you don't need safety (mainly, memory and thread safety,
+among others), then using Rust probably doesn't make sense for you, making this library irrelevant to your needs.
+
+This library follows Rust's [official definition of Safety and Unsafety](https://doc.rust-lang.org/nomicon/meet-safe-and-unsafe.html),
+as detailed in the [Rustonomicon](https://doc.rust-lang.org/nomicon/).
+
+In short, APIs provided by this library that are not marked `unsafe` cannot, under any circumstance, trigger Undefined
+Behavior. This is even more important in the context of real-time processing: incoherent output, or the whole system
+going down, can have terrible consequences.
+
+Safe APIs that can trigger Undefined Behavior are *unsound*, and we consider these bugs: they are either implementation
+bugs or full API design flaws. These are marked with the `Unsound API` tag on GitHub, for better visibility.
+
+**Note:** *While this library is in an alpha (`0.x`) stage, we might allow some soundness issues to stay for a bit, as
+long as they are odd edge-case and not easy to trigger unwillingly. This is because the APIs are not complete
+and still subject to change, however we will have to resolve all of these before getting to `1.0`.*
+
+However, this library can provide APIs that are marked `unsafe`, if there are performance considerations in mind, or if
+those are building blocks for higher levels of abstractions (such as Port Types). In such case, the whole API can be
+made `unsafe`, if appropriate.
+
+The main goal is for end-users to have to write as little `unsafe` code as possible by using this library, ideally none.
+
+This library, however, does not try to be resilient against incorrect LV2 host implementations. Aside from checking
+pointers to be not-null, and checking array indexing to be in-bounds, there is little that a plugin can do. Even when
+a plugin can detect it, there is little it can do to alert the user something is wrong, other than printing to stderr
+and going silent.
+
+Therefore, we need to fully and blindly trust that all the data given to the plugin from the host is correct.
+
+One exception to this rule are the manifest files (the `.ttl` files distributed alongside the plugin binaries). They
+are usually user-written, and writing them incorrectly is likely to trigger Undefined Behavior, in both the plugin and
+the host.
+
+Higher-level, more specialized libraries and frameworks that rely on LV2 should probably auto-generate those files based on
+plugin code and metadata. However, for the low-level control that `lv2` provides, these files need to be user-written
+for now. It is possible that APIs relying on the content of those files could be marked `unsafe` in the future, even
+if that makes it less practical to use for direct users of the `lv2` crates.
+
+## No restrictions coming from C
+
```

Comment:
Ah, yes, I absolutely agree there. I should make this more clear, but this is more of a vision/design goals document, not a roadmap or prioritization document.

This means that while I probably won't push for the less-demanded (or supported) features before a little while (read: some years), I will still accept contributions for them.

Deprecated specifications should probably be implemented at least for hosts (when we support them properly), if they want to be compatible with older plugins. However, I agree this is not really a priority for now.

I agree that ideally, we should only need a viable library before reaching 1.0, and we could add the remaining specs on top of it afterwards. However, I've many times had the experience where adding an extra spec breaks many invariants and assumptions that other specs rely on, and require making breaking changes.

For instance, implementing `Option` requires a refactor of `Atom` to be actually usable, `Instance Access` and `Data Access` break both the single-threadedness and exclusivity invariants of `lv2_core` (basically requiring plugins to be `Sync` instead of just `Send`, and preventing any form of `&mut` reference to ever exist on a plugin instance), and implementing `UI` will probably break *something* that I haven't seen coming yet. :upside_down_face: 

I will add those clarifications to the document. :slightly_smiling_face: